.github/workflows/upgrade-downgrade-e2e.yaml

```diff
@@ -0,0 +1,214 @@
+
+    - name: Create KinD Cluster
+      uses: helm/kind-action@v1.5.0
+      with:
+        config: kind.yaml
+        cluster_name: kind
+        version: v0.17.0
+
+    - name: Get KinD info
+      run: |
+        kubectl cluster-info --context kind-kind
+        NODE_IP=$(kubectl get nodes \
+                  -lkubernetes.io/hostname!=kind-control-plane \
+                  -ojsonpath='{.items[0].status.addresses[?(@.type=="InternalIP")].address}')
+        echo "MINIKUBE_NODE_IP=$NODE_IP" >> $GITHUB_ENV
+
+    - name: Setup test output
+      shell: bash
+      run: |
+        export TEST_OUTPUT_FILE_PREFIX=$GITHUB_WORKSPACE/test_report
+        echo "TEST_OUTPUT_FILE_PREFIX=$TEST_OUTPUT_FILE_PREFIX" >> $GITHUB_ENV
+
+    - name: Setup local registry
+      run: |
+        # Run a registry.
+        docker run -d --restart=always \
+          -p $REGISTRY_PORT:$REGISTRY_PORT --name $REGISTRY_NAME registry:2
+        # Connect the registry to the KinD network.
+        docker network connect "kind" $REGISTRY_NAME
+
+    - name: Setup Helm
+      uses: azure/setup-helm@v1
+      with:
+        version: v3.3.4
+
+    - name: Build and push Dapr
+      run: |
+        make build-linux
+        make docker-build
+        make docker-push
+
+    - name: Build and push test apps
+      run: |
+        cd latest-release && make build-push-e2e-app-all
+
+    - name: Setup Dapr - control-plane master
+      if: matrix.mode == 'control-plane-master'
+      run: |
+        make setup-helm-init
+        make create-test-namespace
+        export ADDITIONAL_HELM_SET="dapr_operator.logLevel=debug,dapr_operator.watchInterval=20s,dapr_sidecar_injector.image.name=ghcr.io/dapr/daprd:${{env.DAPR_LATEST_VERSION}}"
+        export HA_MODE=true
+        make docker-deploy-k8s
+
+    - name: Setup Dapr - control-plane previous release
+      if: matrix.mode == 'dapr-sidecar-master'
+      run: |
+        make setup-helm-init
+        make create-test-namespace
+        export DAPR_CLI_LATEST_VERSION=$(curl -s https://api.github.com/repos/dapr/cli/releases/latest | jq -r '.tag_name' | cut -c 2-)
+        wget -q https://raw.githubusercontent.com/dapr/cli/master/install/install.sh -O - | /bin/bash -s $DAPR_CLI_LATEST_VERSION
+        dapr uninstall --all
+        dapr init -k --runtime-version ${{ env.DAPR_LATEST_VERSION }} \
+          --set dapr_operator.logLevel=debug,dapr_operator.watchInterval=20s,dapr_sidecar_injector.image.name=ghcr.io/dapr/daprd:dev \
+          -n dapr-tests \
+          --enable-ha
+
+    - name: Setup Components
+      run: |
+        cd latest-release
+
+        make setup-test-env-redis
+        make setup-test-env-kafka
+        make setup-test-env-mongodb
+        make setup-test-env-zipkin
+        make setup-test-components
+
+    - name: Free up some diskspace
+      run: |
+        docker image prune -a -f
+
+    - name: Run tests
+      run: |
+        cd latest-release && make test-e2e-all
```

Comment:
Do we need to run the entire test suite? Can we just run a subset of tests that are related to confirming the upgrade works?

Reply:
I've thought about this and originally experimented with running a subset of the current master branch's test suite, adding Go test skips or some other method to run a subset to ensure all appropriate tests would pass. I came to the conclusion that 1. if we run a subset of tests, we will not be testing against the full feature set and will therefore miss incompatibilities, breaking dapr's version skew guarantee, and 2. that guarantee means we _should_ be passing _all_ e2e tests from the previous version, so I see no need not to run them all. It obviously increases test time etc, but not much considering we are already running the full suite 3 times at least once on every PR anyway.

Reply:
Ok, you convinced me :)